File: core/src/main/java/pcb/uwu/amount/derived/electromagnetism/ElectricResistance.java
package pcb.uwu.amount.derived.electromagnetism;

import org.jetbrains.annotations.NotNull;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.CompositeUnitAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.unit.derived.electromagnetism.ElectricResistanceUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.utils.UnitAmountUtils.dividedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.getAmountIn;
import static pcb.uwu.utils.UnitAmountUtils.minusAmount;
import static pcb.uwu.utils.UnitAmountUtils.multipliedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.plusAmount;

public class ElectricResistance extends CompositeUnitAmount<ElectricResistanceUnit> {

	// region constructors

	public ElectricResistance(Number value, ElectricResistanceUnit unit) {
		super(value, unit);
	}

	public ElectricResistance(Number value, Magnitude magnitude, ElectricResistanceUnit unit) {
		super(value, magnitude, unit);
	}

	public ElectricResistance(String value, ElectricResistanceUnit unit) {
		super(value, unit);
	}

	public ElectricResistance(String value, Magnitude magnitude, ElectricResistanceUnit unit) {
		super(value, magnitude, unit);
	}

	public ElectricResistance(BigDecimal value, ElectricResistanceUnit unit) {
		super(value, unit);
	}

	public ElectricResistance(BigDecimal value, Magnitude magnitude, ElectricResistanceUnit unit) {
		super(value, magnitude, unit);
	}

	public ElectricResistance(BigDecimalAmount amount, ElectricResistanceUnit unit) {
		super(amount, unit);
	}

	public ElectricResistance(BigDecimalAmount amount, Magnitude magnitude, ElectricResistanceUnit unit) {
		super(amount, magnitude, unit);
	}

	// endregion

	// region implement UnitAmount

	@NotNull
	@Override
	public ElectricResistance plus(@NotNull UnitAmount<ElectricResistanceUnit> other) {
		return new ElectricResistance(plusAmount(this, other), getUnit());
	}

	@NotNull
	@Override
	public ElectricResistance minus(@NotNull UnitAmount<ElectricResistanceUnit> other) {
		return new ElectricResistance(minusAmount(this, other), getUnit());
	}

	@Override
	public ElectricResistance multiply(BigDecimal other, MathContext mathContext) {
		return new ElectricResistance(multipliedByScalar(this, other, mathContext), getUnit());
	}

	@Override
	public ElectricResistance div(BigDecimal other, MathContext mathContext) {
		return new ElectricResistance(dividedByScalar(this, other, mathContext), getUnit());
	}

	@Override
	public ElectricResistance into(ElectricResistanceUnit unit) {
		return new ElectricResistance(getAmountIn(this, unit), unit);
	}

	// endregion
}

File: core/src/main/java/pcb/uwu/amount/derived/electromagnetism/ElectricResistance.kt
package pcb.uwu.amount.derived.electromagnetism

import pcb.uwu.core.CompositeUnitAmount
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.derived.electromagnetism.ElectricResistanceUnit
import pcb.uwu.utils.UnitAmountUtils

open class ElectricResistance : CompositeUnitAmount<ElectricResistanceUnit>
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL,
                unit: ElectricResistanceUnit)
            : super(amount, magnitude, unit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL,
                unit: ElectricResistanceUnit)
            : super(amount, magnitude, unit)

    // region UnitAmount

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun plus(electricResistance: UnitAmount<ElectricResistanceUnit>) =
        ElectricResistance(amount = this.amount + (electricResistance to this.unit).amount,
                           unit = this.unit)

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun minus(electricResistance: UnitAmount<ElectricResistanceUnit>) =
        ElectricResistance(amount = this.amount - (electricResistance to this.unit).amount,
                           unit = this.unit)

    override fun times(scalar: Number) =
        ElectricResistance(amount = this.amount * scalar,
                           unit = this.unit)

    override fun div(scalar: Number) =
        ElectricResistance(amount = this.amount / scalar,
                           unit = this.unit)

    override fun to(unit: ElectricResistanceUnit) =
        ElectricResistance(amount = UnitAmountUtils.getAmountIn(unitAmount = this, newUnit = unit),
                           unit = unit)

    // endregion
}